library/std/src/os/uefi/env.rs

```diff
@@ -0,0 +1,129 @@
+    NonNull::new(boot_services)
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Get the RuntimeServices Pointer.
+pub fn get_runtime_services() -> Option<NonNull<RuntimeServices>> {
+    let system_table = get_system_table()?;
+    let runtime_services = unsafe { (*system_table.as_ptr()).runtime_services };
+    NonNull::new(runtime_services)
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+/// Open Protocol on a handle
+/// Implemented using `EFI_BOOT_SERVICES.OpenProtocol()`
+pub fn open_protocol<T>(
+    handle: NonNull<c_void>,
+    mut protocol_guid: Guid,
+) -> io::Result<NonNull<T>> {
+    let boot_services = get_boot_services()
+        .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Failed to get BootServices"))?;
+    let system_handle = get_system_handle()
+        .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Failed to get System Handle"))?;
+    let mut protocol: MaybeUninit<*mut T> = MaybeUninit::uninit();
+
+    let r = unsafe {
+        ((*boot_services.as_ptr()).open_protocol)(
+            handle.as_ptr(),
+            &mut protocol_guid,
+            protocol.as_mut_ptr().cast(),
+            system_handle.as_ptr(),
+            crate::ptr::null_mut(),
+            system::OPEN_PROTOCOL_GET_PROTOCOL,
+        )
+    };
+
+    if r.is_error() {
+        Err(super::io::status_to_io_error(r))
+    } else {
+        NonNull::new(unsafe { protocol.assume_init() })
+            .ok_or(io::error::const_io_error!(io::ErrorKind::Other, "Null Protocol"))
+    }
+}
+
+#[unstable(feature = "uefi_std", issue = "none")]
+// Locate handles with a particula protocol Guid
+/// Implemented using `EFI_BOOT_SERVICES.LocateHandles()`
+pub fn locate_handles(mut guid: Guid) -> io::Result<Vec<NonNull<c_void>>> {
+    fn inner(
+        guid: &mut Guid,
+        boot_services: NonNull<BootServices>,
+        buf_size: &mut usize,
+        buf: *mut Handle,
+    ) -> io::Result<()> {
+        let r = unsafe {
+            ((*boot_services.as_ptr()).locate_handle)(
+                r_efi::efi::BY_PROTOCOL,
+                guid,
+                crate::ptr::null_mut(),
+                buf_size,
+                buf,
+            )
+        };
+
+        if r.is_error() { Err(super::io::status_to_io_error(r)) } else { Ok(()) }
+    }
+
+    let boot_services = get_boot_services().ok_or(io::error::const_io_error!(
+        io::ErrorKind::Other,
+        "Unable to acquire boot services"
+    ))?;
+    let mut buf_len = 0usize;
+
+    match inner(&mut guid, boot_services, &mut buf_len, crate::ptr::null_mut()) {
+        Ok(()) => unreachable!(),
+        Err(e) => match e.kind() {
+            io::ErrorKind::FileTooLarge => {}
+            _ => return Err(e),
+        },
+    }
+
+    let mut buf: Vec<Handle> = Vec::with_capacity(buf_len);
+
+    match inner(&mut guid, boot_services, &mut buf_len, buf.as_mut_ptr()) {
+        Ok(()) => {
+            unsafe { buf.set_len(buf_len) };
```

Comment:
Can you add an assertion that the reported necessary buffer size didn't change?

Reply:
Can you explain why that would be needed? There are two cases of buffer size changing:
1. buffer size increases: In this case the second call will return Err of `io::ErrorKind::FileTooLarge`
2. buffer size decreases: In this case, the smaller size will be returned in the `buf_len`, so the set_len will still be correct.

Reply:
> buffer size increases: In this case the second call will return Err of io::ErrorKind::FileTooLarge

I see. I expected it to unconditionally write the used buffer size instead of checking it.

Reply:
Here excerpt from the UEFI specification about the BufferSize field:

> On output, the size in bytes of
the array returned in Buffer (if the buffer was large enough) or the
size, in bytes, of the buffer needed to obtain the array (if the buffer
was not large enough).

In the Status Codes Returned section for this function:

> EFI_BUFFER_TOO_SMALL: The BufferSize is too small for the result. BufferSize has
been updated with the size needed to complete the request.

I will add the comments as well.